libbpf-rs/src/query.rs

```diff
@@ -98,101 +98,305 @@ fn name_arr_to_string(a: &[c_char], default: &str) -> String {
     }
 }
 
+#[derive(Debug, Clone, Default)]
+#[repr(C)]
+struct LineInfoRaw {
+    insn_off: u32,
+    file_name_off: u32,
+    line_off: u32,
+    line_col: u32,
+}
+
+#[derive(Debug)]
+pub struct LineInfo {
+    pub insn_off: u32,
+    pub file_name_off: u32,
+    pub line_off: u32,
+    pub line_num: u32,
+    pub line_col: u32,
+}
+
+impl From<&LineInfoRaw> for LineInfo {
+    fn from(item: &LineInfoRaw) -> Self {
+        LineInfo {
+            insn_off: item.insn_off,
+            file_name_off: item.file_name_off,
+            line_off: item.line_off,
+            line_num: item.line_col >> 10,
+            line_col: item.line_col & 0x3ff,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+#[repr(C)]
+pub struct FuncInfo {
+    insn_off: u32,
+    type_id: u32,
+}
+
+#[derive(Debug, Clone, Default)]
+#[repr(C)]
+pub struct Tag([u8; 8]);
+
 /// Information about a BPF program
 #[derive(Debug)]
 pub struct ProgramInfo {
     pub name: String,
     pub ty: ProgramType,
-    pub tag: [u8; 8],
+    pub tag: Tag,
     pub id: u32,
-    pub jited_prog_len: u32,
```

Comment:
a bit conflicted on removing len fields. They seem useful for cases when user doesn't use ProgInfoQueryOptions and wants to do something smarted and custom on their own after fetching lengths. But this might be a bit hypothethical problem... Having vec fields and separate lens doesn't seem like a big deal?

Reply:
In this model, the _len is just available by calling .len() on the resulting vector. So I do think it's worth it to just do it the clean way in this instance

Reply:
I have no strong feelings either way. Unless we have a good use case for the additional length fields this model of just the `Vec` carrying the length is fine with me. We label ourselves "opinionated eBPF tooling" after all :-)

Reply:
FWIW, upon reflection, the reason to do this would be to write an application that only takes the jited instructions if there are <= N of them (for example). This is likely niche enough to ignore but I'll let people argue with me in V2